R/Factbase.R

```diff
@@ -217,7 +224,25 @@ httrPOST <- function(url=NULL, config=list(), ..., body=NULL, encode=c("multipar
 stop_if_request_too_large <- function (request_body_size) {
     MAX_BODY_SIZE <- 500000000  # matches FUNCTIONS_REQUEST_BODY_SIZE_LIMIT in portal > Factbase > Configuration
     if (request_body_size > MAX_BODY_SIZE)
-        stop(paste0("Your data uses ", request_body_size, "bytes, but the limit is ", MAX_BODY_SIZE), ".  Reduce the quantity of data you are sending.")
+        stop(paste0("Your data uses ", request_body_size, " bytes, but the limit is ", MAX_BODY_SIZE, ".  Reduce the quantity of data you are sending."))
+    if (request_body_size > MAX_BODY_SIZE * .8)
+        warning("Your data uses ", request_body_size, " bytes, which is close to the limit of ", MAX_BODY_SIZE, ".  Consider sending less data.")
+}
+
+# Returning very large data frames for display has two problems:
+# 1. It increases the size of the project file because this output is large,
+#    making the project slow to load/save/undo.
+# 2. In extreme cases the upload will fail because we've breached the maximum
+#    data that can be returned from R calculations.
+truncate_too_large_data <- function (df, max_data_points=1e4) {
+    ncols <- length(df)
+    nrows <- nrow(df)
+    ndata_points <- nrows * ncols
+    if (ncols == 0 || ndata_points <= max_data_points)
+        return(df)
+    message(paste('Truncated data for display, but all', nrows, 'rows have been uploaded to Factbase'))
+    nrows_to_keep <- max_data_points / ncols
```

Comment:
`nrows_to_keep` might not be an integer and I think this will cause problems for the indexing on the next line. Probably use `floor()`.

Reply:
It doesn't cause an error (looks like it is doing floor() internally), but I agree that the code will be more obvious with `floor()`